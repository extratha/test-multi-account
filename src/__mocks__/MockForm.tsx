import { yupResolver } from "@hookform/resolvers/yup";
import { ReactNode, useEffect } from "react";
import { FormProvider, UseFormReturn, useForm } from "react-hook-form";
import { ObjectSchema } from "yup";

export interface MockFormProps<T = any> {
  initValues: T | undefined;
  resolver?: ObjectSchema<any>;
  children?: ReactNode;
  onInitForm?: (methods: UseFormReturn) => void;
  onSubmit: () => void;
}

export interface MockHandleValueProps {
  name: string;
}

const MockForm = (props: MockFormProps) => {
  const { initValues, resolver, children, onInitForm, onSubmit } = props;
  const methods = useForm({
    reValidateMode: "onSubmit",
    defaultValues: initValues,
    resolver: resolver ? yupResolver(resolver) : undefined,
  });

  useEffect(() => {
    if (onInitForm) onInitForm(methods);
  }, []);

  return (
    <FormProvider {...methods}>
      <form noValidate onSubmit={methods.handleSubmit(onSubmit)}>
        {children}
        <button type="submit" data-testid="submit-button">
          Submit
        </button>
      </form>
    </FormProvider>
  );
};

export default MockForm;
