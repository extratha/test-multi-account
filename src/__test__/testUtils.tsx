import { RenderOptions, RenderResult, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { SWRConfig } from 'swr';
import { theme } from '@/config/config-mui/theme';
import { ThemeProvider as MUIThemeProvider } from '@mui/material/styles';
import { NextIntlClientProvider } from 'next-intl';
import commonTh from '../../public/locales/th/common.json';
import commonEn from '../../public/locales/th/common.json';
type RenderWithProvider = (
  elm: React.ReactElement,
  renderOptions?: ProviderOptions,
) => RenderResult;

export type ProviderOptions = {
  formOptions?: Record<string, unknown>;
} & RenderOptions;

const messages = {
  ...commonTh,
  ...commonEn,
};

const renderWithProviders: RenderWithProvider = (
  component,
  renderOptions?,
  locale = 'th',
) => {
  const wrapper = ({ children }: { children: React.ReactNode }) => (
    <NextIntlClientProvider locale={locale} messages={messages}>
      <MUIThemeProvider theme={theme}>
        <SWRConfig value={{ provider: () => new Map() }}>{children}</SWRConfig>
      </MUIThemeProvider>
    </NextIntlClientProvider>
  );

  return render(component, {
    wrapper,
    ...renderOptions,
  });
};

export { renderWithProviders as render, userEvent };
export * from '@testing-library/react';
