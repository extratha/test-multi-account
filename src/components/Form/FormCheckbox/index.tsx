// import { CUSTOM_COLORS } from "@/config/config-mui/theme/colors";
import { Checkbox, FormControlLabel, Typography } from "@mui/material";
import { useController, useFormContext } from "react-hook-form";

export interface FormCheckboxProps {
  name: string;
  label?: string;
  disabled?: boolean;
  className?: string;
}

const FormCheckbox = ({ name, label, disabled, className }: FormCheckboxProps) => {
  const { control } = useFormContext();
  const { field } = useController({ name, control });

  return (
    <FormControlLabel
      disabled={disabled}
      data-testid={`${name}-checkbox`}
      control={
        <Checkbox
          {...field}
          checked={field.value}
          data-testid={`${name}-checkbox-item-button`}
          // @ts-ignore
          inputProps={{ "data-testid": `${name}-checkbox-value` }}
        />
      }
      label={<Typography data-testid={`${name}-checkbox-label`}>{label}</Typography>}
      className={className}
    />
  );
};

export default FormCheckbox;
