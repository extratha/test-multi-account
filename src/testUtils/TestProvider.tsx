import { ThemeProvider } from "@emotion/react";

import { ReactNode } from "react";
import { SWRConfig } from "swr";

import ModalLayer from "@/components/Modal/ModalLayer";
import { theme } from "@/config/config-mui";

interface TestProviderProps {
  children?: ReactNode;
}

export const TestProvider = ({ children }: TestProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <SWRConfig value={{ provider: () => new Map() }}>{children}</SWRConfig>
      <ModalLayer />
    </ThemeProvider>
  );
};
