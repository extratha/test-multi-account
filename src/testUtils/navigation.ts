import * as NextNavigation from "next/navigation";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
  useParams: jest.fn(),
  usePathname: jest.fn(),
  useSearchParams: jest.fn(),
  useServerInsertedHTML: jest.fn(),
}));

export interface SpyUseSearchParams {
  get: jest.Mock;
}

export interface SpyUseRouter {
  replace: jest.Mock;
  push: jest.Mock;
  back: jest.Mock;
}

export const spyUseParams = (params: any = {}) => {
  jest.spyOn(NextNavigation, "useParams").mockReturnValue(params);
  return NextNavigation.useParams as jest.Mock;
};

export const spyUseRouter = (): SpyUseRouter => {
  const replace = jest.fn();
  const push = jest.fn();
  const back = jest.fn();
  jest.spyOn(NextNavigation, "useRouter").mockReturnValue({ replace, push, back } as any);
  return { replace, push, back };
};

export const spyUsePathname = (pathname: string) => {
  jest.spyOn(NextNavigation, "usePathname").mockReturnValue(pathname);
};

export const spyUseSearchParams = (): SpyUseSearchParams => {
  const get = jest.fn();
  jest.spyOn(NextNavigation, "useSearchParams").mockReturnValue({ get } as any);
  return {
    get,
  };
};

export const spyUseServerInsertedHTML = () => {
  return jest.spyOn(NextNavigation, "useServerInsertedHTML");
};
