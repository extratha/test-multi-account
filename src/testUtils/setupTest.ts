import "@testing-library/jest-dom";
import "jest-canvas-mock";

import "@/__mocks__/Markdown";
import "@/__mocks__/MockFirebase";
import "@/__mocks__/NextImage";
import { LOCALE_LANGUAGE } from "@/config";
import { initI18next } from "@/locales/i18n";
import locale from "@/locales/th.json";

initI18next(LOCALE_LANGUAGE.TH, locale);

jest.mock("@/api/api", () => {
  const actualModule = jest.requireActual("@/api/api");
  return { ...actualModule, __esModule: true };
});

jest.mock("@/api/apiUnauthorize", () => {
  const actualModule = jest.requireActual("@/api/apiUnauthorize");
  return { ...actualModule, __esModule: true };
});

jest.mock("next/config", () => () => ({
  setConfig: jest.fn(),
  publicRuntimeConfig: {
    apiToken:
      "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE3MDY2NzUzNDQsImlhdCI6MTcwNjU4ODk0NCwicHJvZmlsZUlkIjoiZDU1YWY3NDUtOTYyZS00MTk2LTkyOTEtZDg2YTliYWNiZWY2In0.sNgwYjIGLYEVAptqILGgRVGXwt9FCPKHRubEZyaodhg",
    mockBaseApiUrl: "https://www.mock-playground-dashboard.co.th",
    baseApiUrl: "https://www.mock-playground-dashboard.co.th",
    globalEndpoint: "playground-dashboard",
    publicApiHost: "http://localhost:3000/",
  },
}));

jest.mock("cookies-next", () => ({
  __esModule: true,
  ...jest.requireActual("cookies-next"),
  getCookie: jest.fn().mockReturnValue("MOCK_ACCESS_TOKEN"),
}));
