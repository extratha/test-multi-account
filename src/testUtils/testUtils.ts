import { act, render as testRender } from "@testing-library/react";
import dayjs from "dayjs";
import flushPromises from "flush-promises";
import { ReactElement } from "react";

import { TestProvider } from "./TestProvider";

export const API = {
  LOGIN: "/auth/login",
  EXAMPLE_AI_INTERPRET_URL: "/lab/examples/labId",
  GET_LAB_INTERPRETS_ID_URL: "/lab/interprets/labId",
  SUBMIT_LAB_INTERPRETS_URL: "/lab/interprets",
  CONSENT: "/consents",
  TERMS_AND_CONDITIONS: "/consents/terms-conditions",
  TERMS_AND_CONDITIONS_LATEST: "/consents/terms-conditions/latest",
  PRIVACY_POLICY: "/consents/privacy-policies",
  PRIVACY_POLICY_LATEST: "/consents/privacy-policies/latest",
  PUBLIC_TERMS_CONDITIONS: "/consents/terms-conditions/public",
  PUBLIC_PRIVACY_POLICIES: "/consents/privacy-policies/public",
};

export const render = (ui: ReactElement) => {
  return testRender(ui, { wrapper: TestProvider });
};

export const flushPromise = async () => {
  await act(async () => {
    await flushPromises();
  });
};

export const advanceTimersByTime = async (time: number) => {
  await act(async () => {
    jest.advanceTimersByTime(time);
  });
};

export const setMockDate = (date: string) => {
  jest.setSystemTime(dayjs(date).valueOf());
};
