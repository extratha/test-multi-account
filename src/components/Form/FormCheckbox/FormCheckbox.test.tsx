import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import * as yup from "yup";

import MockForm from "@/__mocks__/MockForm";
import FormCheckbox, { FormCheckboxProps } from "@/components/Form/FormCheckbox";
import { render } from "@/testUtils/testUtils";

describe("FormCheckbox", () => {
  let props: FormCheckboxProps;
  let form: any;

  beforeEach(() => {
    props = {
      name: "test",
      label: "test label",
    };

    form = {
      initValues: { test: false },
      resolver: yup.object().shape({
        test: yup.boolean().required("Message Error"),
      }),
      onSubmit: jest.fn(),
    };
  });

  const renderComponent = () => {
    return render(
      <MockForm {...form}>
        <FormCheckbox {...props} />
      </MockForm>
    );
  };

  it("should render correctly", () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should check/uncheck when click checkbox", async () => {
    renderComponent();

    expect(screen.getByTestId("test-checkbox-value")).not.toBeChecked();

    await userEvent.click(screen.getByTestId("test-checkbox"));
    expect(screen.getByTestId("test-checkbox-value")).toBeChecked();

    await userEvent.click(screen.getByTestId("test-checkbox"));
    expect(screen.getByTestId("test-checkbox-value")).not.toBeChecked();
  });
});
