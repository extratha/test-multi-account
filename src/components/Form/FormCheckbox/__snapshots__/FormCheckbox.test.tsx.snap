// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckbox should render correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  vertical-align: middle;
  -webkit-tap-highlight-color: transparent;
  margin-left: -11px;
  margin-right: 16px;
}

.emotion-0.Mui-disabled {
  cursor: default;
}

.emotion-0 .MuiFormControlLabel-label.Mui-disabled {
  color: #B0BEC5;
}

.emotion-1 {
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 50%;
  padding: 9px;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: rgba(0, 0, 0, 0.6);
}

.emotion-1::-moz-focus-inner {
  border-style: none;
}

.emotion-1.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-1 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-1:hover {
  background-color: rgba(33, 150, 243, 0.04);
}

@media (hover: none) {
  .emotion-1:hover {
    background-color: transparent;
  }
}

.emotion-1.Mui-checked,
.emotion-1.MuiCheckbox-indeterminate {
  color: #2196F3;
}

.emotion-1.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
}

.emotion-2 {
  cursor: inherit;
  position: absolute;
  opacity: 0;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  margin: 0;
  padding: 0;
  z-index: 1;
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
}

.emotion-4 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

.emotion-5 {
  margin: 0;
  font-family: fontFamily;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
}

<form
    novalidate=""
  >
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd emotion-0"
      data-testid="test-checkbox"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium emotion-1"
        data-testid="test-checkbox-item-button"
      >
        <input
          class="PrivateSwitchBase-input emotion-2"
          data-indeterminate="false"
          data-testid="test-checkbox-value"
          name="test"
          type="checkbox"
          value="false"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium emotion-3"
          data-testid="CheckBoxOutlineBlankIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root emotion-4"
        />
      </span>
      <p
        class="MuiTypography-root MuiTypography-body1 emotion-5"
        data-testid="test-checkbox-label"
      >
        test label
      </p>
    </label>
    <button
      data-testid="submit-button"
      type="submit"
    >
      Submit
    </button>
  </form>
</DocumentFragment>
`;
